internal/wasm/gofunc.go

```diff
@@ -46,6 +47,94 @@ func GetHostFunctionCallContextValue(fk FunctionKind, ctx *ModuleContext) *refle
 	return nil
 }
 
+// PopGoFuncParams pops the correct number of parameters off the stack into a parameter slice for use in CallGoFunc
+//
+// For example, if the host function F requires the (x1 uint32, x2 float32) parameters, and
+// the stack is [..., A, B], then the function is called as F(A, B) where A and B are interpreted
+// as uint32 and float32 respectively.
+func PopGoFuncParams(f *FunctionInstance, popParam func() uint64) []uint64 {
+	// First, determine how many values we need to pop
+	paramCount := f.GoFunc.Type().NumIn()
+	if f.Kind != FunctionKindGoNoContext {
+		paramCount--
+	}
+
+	return PopValues(paramCount, popParam)
+}
+
+// PopValues pops api.ValueType values from the stack and returns them in reverse order.
+//
+// Note: the popper intentionally doesn't return bool or error because the caller's stack depth is trusted.
+func PopValues(count int, popper func() uint64) []uint64 {
+	params := make([]uint64, count)
+	for i := count - 1; i >= 0; i-- {
+		params[i] = popper()
+	}
+	return params
+}
+
+// CallGoFunc executes the FunctionInstance.GoFunc by converting params to Go types. The results of the function call
+// are converted back to api.ValueType.
+//
+// * callCtx is passed to the host function as a first argument.
+//
+// Note: ctx must use the caller's memory, which might be different from the defining module on an imported function.
+func CallGoFunc(callCtx *CallContext, f *FunctionInstance, params []uint64) []uint64 {
```

Comment:
this is the new centralized function. Before, it was copy/pasted for JIT and Interpreter. While not doing anything too fancy now, you can imagine we can extract a template-based call invoker here one day if we wanted.